sway-ir/src/analysis/memory_utils.rs

```diff
@@ -45,16 +45,121 @@ impl Symbol {
+    /// Denotes that there _might_ be [Symbol]s out of returned ones that
+    /// are related to the particular [Value], but not reachable only via GEP,
+    /// or references, or raw pointers.
+    Incomplete(FxIndexSet<Symbol>),
+}
+
+impl ReferredSymbols {
+    // TODO: Check all the usages of this method and replace it with the
+    //       checked access to either complete or incomplete symbols.
+    //       This is a temporary convenience method until
+    //       we decide case by case how to deal with incomplete set of symbols.
+    //       See: https://github.com/FuelLabs/sway/issues/5924
+    pub fn any(self) -> FxIndexSet<Symbol> {
+        match self {
+            ReferredSymbols::Complete(symbols) | ReferredSymbols::Incomplete(symbols) => symbols,
+        }
+    }
+}
+
+/// Get [Symbol]s, both [Symbol::Local]s and [Symbol::Arg]s, reachable
+/// from the `val` via chain of [InstOp::GetElemPtr] (GEP) instructions
+/// or via [InstOp::IntToPtr] and [InstOp::PtrToInt] instruction patterns
+/// specific to references, both referencing (`&`) and dereferencing (`*`),
+/// and raw pointers, via `__addr_of`.
+/// A `val` can, via these instructions, refer indirectly to none, or one
+/// or more symbols.
+///
+/// Note that *this function does not perform any escape analysis*. E.g., if a
+/// local symbol gets passed by `raw_ptr` or `&T` to a function and returned
+/// back from the function via the same `raw_ptr` or `&T` the value returned
+/// from the function will not be tracked back to the original symbol and the
+/// symbol will not be collected as referred.
+///
+/// This means that, even if the result contains [Symbol]s, it _might_ be that
+/// there are still other [Symbol]s in scope related to the `val`. E.g., in case
+/// of branching, where the first branch directly returns `& local_var_a`
+/// and the second branch, indirectly over a function call as explained above,
+/// `& local_var_b`, only the `local_var_a` will be returned as a result.
+///
+/// Therefore, the function returns the [ReferredSymbols] enum to denote
+/// if the returned set of symbols is guaranteed to be complete, or if it is
+/// incomplete.
+pub fn get_referred_symbols(context: &Context, val: Value) -> ReferredSymbols {
+    get_symbols(context, val, false)
+}
+
+/// Get [Symbol]s, both [Symbol::Local]s and [Symbol::Arg]s, reachable
+/// from the `val`.
+///
+/// If `gep_only` is `true` only the [Symbol]s reachable via GEP instructions
+/// are returned. Otherwise, the result also contains [Symbol]s reachable
+/// via referencing (`&`) and dereferencing (`*`).
+fn get_symbols(context: &Context, val: Value, gep_only: bool) -> ReferredSymbols {
     let mut visited = FxHashSet::default();
     let mut symbols = IndexSet::default();
     fn get_symbols_rec(
         context: &Context,
         symbols: &mut FxIndexSet<Symbol>,
         visited: &mut FxHashSet<Value>,
         val: Value,
+        gep_only: bool,
+        is_complete: &mut bool,
     ) {
```

Comment:
Actually, it's not the ownership but the fact that in Rust variables from the outer function scope are not visible in nested functions.

> If that is the case I would suggest moving the variable declarations of visited and symbols to after the declaration of the helper functions, so that they sit alongside the declaration of is_complete.

Ha ha actually I wanted to do that but didn't want to change the existing code with my own preference. But now it's two of us finding it a bit confusing having those two declarations on the top. Also, after including `is_complete` they should all be either on the top or on the bottom. So, as suggested, moving them all down.